src/compiler/checker.ts

```diff
@@ -8964,6 +8964,13 @@ namespace ts {
                         (isTypeSubsetOf(globalObjectType, target) || (!isComparingJsxAttributes && isEmptyObjectType(target)))) {
                         return false;
                     }
+                    if (target.flags & TypeFlags.Union) {
+                        const discriminantType = findMatchingDiscriminantType(source, target as UnionType);
+                        if (discriminantType)  {
+                            // check excess properties against discriminant type only, not the entire union
+                            return hasExcessProperties(source, discriminantType, reportErrors);
+                        }
+                    }

```

Comment:
This is similar to the problem with #14006, which added `findMatchingDiscriminantType`, although there the aim was to improve an existing error, so it was still improvement to return one of multiple possible types with matching discriminants.

I think a fix here would be to make findMatchingDiscriminantType not return anything if multiple types match. However, I'm not sure what type shows this behaviour. Is it something like `{ tag: "A", x } | { tag: "A", y }` ?